### ISSUE TITLE
Auto-Accept shared folders from trusted devices

### ISSUE THREAD

**Issue by Getron** [syncthing/syncthing#2299]
This is a proposal of auto-accepting shared folders from a trusted devices.

For example, if i want to add more and more shares on my servers, i must on every server to accept the shares, every share, on every device.

If the share come from the trusted (added) device, maybe to have an option on device, to check a box, if the device is trusted (or own), and to give an standard folder location on the server or machine, where the auto-accepted shares would go (maybe, if the auto-accepted share is on, then, on this standard folder, the syncthing maybe create an subfolder with the name of shared folder, to have it sorted, if there are many shared folders), and would be accepted automaticaly.

Like this, user must not manualy accept every share, on every sync machines (devices), separatelly. Just add the shares on local machine, or local server, and have it done on all other machines (devices, or servers).

Devs, what you think? Is this an good idea?

**Comment by marrie** [comment]
I agree. I've been having a  similar discussion on the forums but about a queue to except devices if you don't have the web UI up, maybe to auto except devices as well. A lot of people think the idea sucks, but I personally think both your idea and mine would make for a better work flow. Of corse these could be toggles for the user who does not want the option.

**Comment by AudriusButkevicius** [comment]
The question for the original post, would the location where folders get created would be per device or global? Also, what happens if a folder under the same name already exists? Do we share the existing one?

**Comment by AudriusButkevicius** [comment]
As for devices, I don't think it will happen as adding a device does not share any folders hence adds no value. And adding new device and sharing folders automatically is just plain damn dangerous.

**Comment by Getron** [comment]
I think, just auto-add the share from "Trusted" devices.

Maybe, to add an option on the device adding, a check aquare, to mark the device as own, or trusted. Not the devices itself to be added. Just shares from a trusted devices. This is safe.

Example:
I have 2 devices: 1 PC and one Server

I share a folder from my PC with server. On the server side, i should just once add the PC, and on adding (or later), on the device options, i should able to mark the PC device as a trusted one.

All future shared folders from my PC should automaticaly added on my server.

The location where the folder where created (in my example, server), i should maybe choose on the main options of syncthing. Or maybe in the "Sync" folder itself to add a folder "Auto-added"(this is a standard one), but, it should be great, if the user can choose one.

After this, on this main folder, if, for example, my PC syncthing name is "Main-PC", on the server side, the syncthing should create a folder with the name of devices, on the main folder.

It should look like this:

Sync/Auto-added/Main-PC/   and here, for example, my shared folder is Doku, should add this Doku shared folder there.

And yes, if the same name from same trusted device, we share this, as synthing it does now.

**Comment by calmh** [comment]
I think it would be worth to take the time to specify all the various cases that may occur here before sitting down to implement. (But I agree this is probably worth doing.) As in, what happens when

 - Trusted device pops in, wants to share a new folder
 - Trusted device pops in, wants to share an existing folder
 - Trusted device pops in, wants to share a new folder, then another device comes along and wants to share the same folder
 - etc

**Comment by Getron** [comment]
I think, we should not change to much here, my proposal is:

- Trusted device pops in, wants to share a new folder - the folder should be added on the main location for trusted shares (given as default on main syncthing options), or if user give an specified location for trusted device (on device options), create this folder there.
- Trusted device pops in, wants to share an existing folder - folder should be shared like it is now shared. If exist, share it, if not, create it.
- Trusted device pops in, wants to share a new folder, then another device comes along and wants to share the same folder - folder should be created, and shared with this new devices, just in case, if this devices are trusted one on the main Trusted device who pops. Other wise, it is safe to not share trusted stuff with, maybe untrusted devices.

**Comment by marrie** [comment]
I like the idea of that workflow.  if the devices share the same folder, or if a new device shares a folder with thesame name, there shoudl be an option to merge, replace, overwrite or rename the new folder so newfolder-1 for example.

**Comment by Getron** [comment]
Yeah, that's more better, if let the user decide what should happend with shares of Trusted devices. Maybe this options should we have on every share, to decide if merge, replace, overwrite or rename the existing folder.

**Comment by AudriusButkevicius** [comment]
This is becoming too many bells and whistles.

**Comment by uok** [comment]
related - https://github.com/syncthing/syncthing/issues/2157

**Comment by Getron** [comment]
@AudriusButkevicius I known, it is not easy to do all the stuff, and it is huge work, but, my idea is to have an radio button on device options, to select if this device is Trusted, and if is, then, add the shares from this device on given location (default, or selected one - this field should be changeable, or selectable, just if user click on the Trusted device button)

On the ask of @calmh , what if ... points... then, should ask the user, what to do, the 4 options (merge, replace, overwrite or rename).

And that's all. (nothing complicated for every user, if the software ask him, what to do, if...)

Like this, for example, if user have an servers around the world, it is not needed to access the servers with teamviewer, or somthing like this, and to accept every share - with this option, user need just once to add the device as Trusted, set the location, and what should happend, if ( @calmh points) - done.

Syncthing become more automatised, and more usable. Just think, if user share a couple of shares, the full internet bandwidth is used, and i need to connect with an remote desktop like teamviewer, or No Machine, the work on this remote machine is extremly slow and unusable.

With this option, user doesn't need remote desktop software anymore, to access the servers, or machine around the world, just for accept the new shares.

**Comment by ysmolsky** [comment]
After reading full discussion, I do not think that this ticket is easy in any way :) Should it have "easy" label in this case?

**Comment by AudriusButkevicius** [comment]
I think the naive implementation is still easy.

**Comment by Getron** [comment]
@AudriusButkevicius my ideas, or proposals are not my "bells an whistless", Sorry.

This are just proposals, that, i think, it would help many other peoples, and, if we have an automated software (Syncthing), why should it not make better, and even more easy and automated, for business directions to (like for Server farms, and Hostings people), not just for private use. Ok, it can be used on business directions to, but, every hoster, who do the Backups, and such things with Syncthing, would apreciate the automation to.

Like the version 0.12.x with pausing devices, improved refreshing time (allmost no pause time on refreshing now, wonderfull), and all of this ideas are realised, and every person now is happy to have this, it works wonderfull. Maybe, not everyone need the pausing devices, for example, but, the function is there, and "if" someone need it, it is there - and this is wonderfull, it works great.

Unfortunatly, i'm not so good on programming, and can not help as much as i wish. Therefore, i give just ideas, and proposals, and this, are not "bells an whistless", this are just proposals how it maybe works at best the proposal.

Anyway, would thank an all of devs for time spended for having a wonderfull software.

**Comment by kluppy** [comment]
I think this is a duplicate of #1839.
I want like the idea and would like to see it implemented.
Perhaps I will give it a go once I try my hand at GO.

**Comment by gaborgabor** [comment]
I really need this function

**Comment by calmh** [comment]
Depends on #2157

**Comment by user514** [comment]
I have bittorrent sync and syncthing installed on my raspberry2. Was just looking for the exact same function Getron describes in the first post here. I found that bittorrent does that when you are a 'Pro' user, then you can sync across all your devices from just one client by selecting in which computers your new folder should appear and be auto accepted (seems the main difference between free and paid version).

  I'm just writing here to subscribe the Getron proposal. I think that feature for free on syncthing will catch the interest of many people that have to go for the commercial bittorrent because of this advantage.

**Comment by Getron** [comment]
@user514 Thank you for understanding my proposal. It should be a extremly great feature, if implemented. Syncthing is an automated sync. but, why not make it even more easy and automated? I think, this would be a big interest for private, and even more bussines people.

Thanks to devs in advance, if this would be implemented, it should help to menage syncing to any servers, without have to remotelly screening every time new folder is added.

Should work somthing like is the way that introducer device system works, if selected as trusted, all devices from this device add itself automaticaly.

Exatly the same should be option with the folder (maybe with a few options there, i described earlyer posts).

**Comment by urza** [comment]
Hi, I just wanted to share my interest in this feature. The auto-accept from trusted devices.

I would say, 
1. set default path for new auto-shared folders
2. mark/set devices that are trusted and can auto-share folders with me
3. name those new folders something like auto-shared-fromDevice-ForlderID or something like that...

**Comment by spectras** [comment]
…and don't forget to allow disabling the feature altogether.

**Comment by cherylquirion** [comment]
I have done a preliminary pass to provide this functionality (with limitations). More info here: https://github.com/syncthing/syncthing/pull/3892

**Comment by LosEagle** [comment]
So I guess this is not happening? I'd like to see this feature as well since it can be quite a pain when trying to sync individual folders with device that is miles away from the central node. Not even mentioning that both devices have to be connected when one accepts permission to sync on central node. This can lead to having to keep device turned on for hours/days just to sync one folder.

**Comment by spectras** [comment]
It is not necessary, and it will never be. Simply use proper configuration management tools to manage your configuration ([ansible](https://www.ansible.com/) comes to mind but there are plenty others) and you'll find auto accepting not only is useless with a proper setup, but would actually lower the trust you can have in your solution (this is the reason for which, should that ever be implemented, the option to disable it entirely would be essential).

Now I understand it would be *convenient*, as in *it gets the job done without having to design a configuration management strategy*.
